cty/msgpack/marshal.go

```diff
@@ -32,7 +32,10 @@ func Marshal(val cty.Value, ty cty.Type) ([]byte, error) {
 	var buf bytes.Buffer
 	enc := msgpack.NewEncoder(&buf)
 	enc.UseCompactInts(true)
-	enc.UseCompactFloats(true)
+
+	// UseCompactFloats can fail on some platforms due to undefined behavior of
+	// float conversions
+	enc.UseCompactFloats(false)
```

Comment:
Just a note for the future, since I went and looked this all up to remind myself for review purposes...

---

`UseCompactFloats` here means "encode floats as integers when possible", which can potentially be productive because MessagePack has some specialized compact representations of small integers, whereas floats are always encoded as either 5 bytes (for `float32`) or 9 bytes (for `float64`).

That optimization seemed useful because `cty.Number` doesn't distinguish between integers and floats anyway -- it's just "the number type" -- and so we'd prefer to encode all numbers in the most compact form possible.

However, the problem arises in how the upstream library makes the decision about whether to do it. It uses `float64(int64(n)) == n` (where `n` is `float64`), assuming that this will return true if and only if `n` is a float representation of an integer. That doesn't hold true on all architectures, which is a specific allowance made in the current Go specification:

> In all non-constant conversions involving floating-point or complex values, if the result type cannot represent the value the conversion succeeds but the result value is implementation-dependent.

It turns out that on some architectures this causes the expression above to return `true` even when `n` is not an integer if `n` is too large to fit into `int64`. That then causes the `cty` MessagePack encoding to silently lose precision when round-tripping.

---

This change doesn't actually cause any serialization bloat for valid cases in practice anyway, because the code in _this_ package that calls into `enc` methods has its own special case for treating whole numbers as integers, using the `math/big.Float.Int64` definition instead of relying on the CPU's built-in type conversion rules:

https://github.com/zclconf/go-cty/blob/1348b6911c5e75cb237d6e58168f1dd390666349/cty/msgpack/marshal.go#L86-L93

Because we're still setting `UseCompactInts`, `enc.EncodeInt` will still select the most compact integer representation. But now when we call `enc.EncodeFloat64` we will definitely always get the 9-byte MessagePack float64 representation.